src/beatree/leaf/store.rs

```diff
@@ -150,53 +125,59 @@ impl<'a> LeafStoreTx<'a> {
             }
         };
 
-        if leaf_pn.0 < self.max_bump.0 {
-            self.to_allocate.push((leaf_pn, leaf_page));
-        } else {
-            self.exceeded.push((leaf_pn, leaf_page));
-        }
+        self.pending.push((leaf_pn, leaf_page));
+
         leaf_pn
     }
 
     pub fn release(&mut self, id: PageNumber) {
         self.released.push(id);
     }
 
-    // commits the changes creating a set of pages that needs to be written into the LeafStore
-    pub fn commit(mut self) -> LeafStoreTxOutput {
-        let free_list_output = self
-            .free_list
-            .commit(self.released, &mut self.bump, self.max_bump);
+    // Commits the changes creating a set of pages that needs to be written into the LeafStore.
+    //
+    // The output will include not only the list of pages that need to be written but also
+    // the new free_list head, the current bump page number, and the new required file size
+    pub fn commit(&mut self) -> LeafStoreCommitOutput {
+        let released = std::mem::take(&mut self.released);
+        let pending = std::mem::take(&mut self.pending);
 
-        let to_allocate = self
-            .to_allocate
-            .into_iter()
-            .map(|(pn, leaf_page)| (pn, leaf_page.inner))
-            .chain(
-                free_list_output
-                    .to_allocate
-                    .into_iter()
-                    .map(|(pn, free_list_page)| (pn, free_list_page.inner)),
-            )
-            .collect::<Vec<_>>();
+        let free_list_pages = self.free_list.commit(released, &mut self.bump);
 
-        let exceeded = self
-            .exceeded
+        let pages = pending
             .into_iter()
             .map(|(pn, leaf_page)| (pn, leaf_page.inner))
             .chain(
-                free_list_output
-                    .exceeded
+                free_list_pages
                     .into_iter()
                     .map(|(pn, free_list_page)| (pn, free_list_page.inner)),
             )
             .collect::<Vec<_>>();
 
-        LeafStoreTxOutput {
-            to_allocate,
-            exceeded,
+        // The LeafStore is expected to grow in multiples of 1 MiB blocks,
+        // so even if the self.bump exceeds the maximum supported bump by 2,
+        // the file will be resized to store an additional 256 pages
```

Comment:
Nit: once again, "maximum supported bump" I find v. confusing.
But why "by 2"?

Reply:
docs updated! "by 2" was just an example to demonstrate how there will possible be some empty pages at the end of the file that will not being used by the pages produced by the commit